### ISSUE TITLE
Proposal: Resolve method call with duplicate interface method definitions

### ISSUE THREAD

**Issue by UpperAvenue** [dotnet/roslyn#46936]
### Background

Consider an interface with multiple generic type parameters and that defines overloaded methods that differ only by the generic type used as the type of a parameter.   When the same type is used as the generic type arguments, these overloaded methods definitions are essentially _duplicated_.  These duplicated methods definitions can only be implemented by a single method, in an implementing class.  

Example:

**Interface:**
```csharp
interface IMap<T,U> 
{
	bool Contains(T val);
	bool Contains(U val);
}
```

**Implementation:** 
```csharp
class Map<T> : IMap<T,T> 
{
	public bool Contains(T val) => true;
}
```

Currently, when calling Contains on a IMap<T,T>, the error CS0121 for ambiguous method call is raised.
```csharp
IMap<int,int> map = new Map<int>();
...
map.Contains(42); // CS0121
``` 

### Proposal
Since these duplicate interface method definitions can only resolve to a single method, the calling method should resolve to the last defined method that matches in the interface.

**Comment by svick** [comment]
> These duplicated methods definitions can only be implemented by a single method, in an implementing class.

That's not true, consider [the following generic implementation](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ7EAI5IOzoDe6WpOSAbDgCxYCyAhgJZQAUAlCWcWmX1gElGABwA8LYABoJAPiwBbBsKwBeLFACmAd3pLxUKVlkcA3F35YA9JYVKAdAGEA9geZQAzq2oJ2JqzfdgJgAbYKwHAGUABiRkc354vjhaOlEAFUksAFUZViE9DOy5RWFOXgssHgr+XABOVhLHF2A3TwATDQAzBghg4FY09iG/ayxgSA1EirqG+2dXFnaunr7WLKHffyxu4PdJ8osAXynSE/pZ0rMDrGO0W/QJDTBugGMNQRF0yRysInQASAARk4nGF5i1FgMsAA3BjBXwA4Gg8LNVprGFwhF3dCYBC6MSFH4gD4FTI5P7XOAAZiwSLBqMhaQx8NUcnGEA0VySNLpKIWHnRsJZKjkOz2V0OQA=):

```c#
class Map<T, U> : IMap<T, U>
{
    public bool Contains(T val) => true;
    public bool Contains(U val) => false;
}
```

Usage:

```c#
IMap<int,int> map = new Map<int, int>();
// map.Contains(42); // still CS0121

void M<T, U>(IMap<T, U> map)
{
    Console.WriteLine(map.Contains(default(T))); // true
    Console.WriteLine(map.Contains(default(U))); // false
}

M(map);
```

**Comment by UpperAvenue** [comment]
@svick in your case the implementing class uses two different generic types, T and U for IMap<T,U> with the instantiated _object_ having the same types. What I’m saying is that when the implementing classes uses the same generic type arguments for IMap it should be able to resolve. 

 Examples:
```csharp 
class IntMap : IMap<int,int> { … }
class SingleTypeMap<T> : IMap<T,T> { … }
class BaseListAndDerivedMap<T, U> : IEnumerable<T>, IMap<U,U> where U : T
```

**Comment by svick** [comment]
@UpperAvenue In your case, the type of the problematic variable was `IMap<int, int>`. How is the compiler supposed to figure out whether the implementation has a single method implementing the two interface methods or not?

**Comment by UpperAvenue** [comment]
@svik would it not be able to figure it out if `T` and `U` of `IMap<T,U>` are the same type.  If it is `IMap<int,int>` we know that `Contains(T val)` and `Contains(U val)` can only be implemented by a singular method `Contains(int val)`.  

[SharpLab example](https://sharplab.io/#v2:C4LgTgrgdgPgAgBgARwIwBYDcBYAUIlVAOgGEB7AGwoFMBjYASzKgGcdc84AmJAMTLJIA3gF9OPAEIBDMEhB8BwsR3w80AdjxC8SXYQBsKdEgCyUhlAAUASh17tuPU6QBJMwAcAPPzIAaHwB8SAC2Uu5IALxIUNQA7qZh3gIBNuzOem6JFsC+2UHZHpHRca5QwB6pdumZXtJgvnVBtGTB7jQAHoVRMfHSLNQAMgwswACCUAAmACLUYAwAbtQTHkl+SI2VjulIVc67TmiGcMYmngAqvkgAqik155c3IWG2W+kO286h7qTMwOaslgm1AAZlIIBRgJYztZrGkPnovj8yv8WICQWCIZYrjC4ellB99noTJYvrDCbpiQVnrjnMTmq0Oh4ya89izlMo8NlZqDaNRXCsLo8tHgAJAAIwEFCQ5GRFlRZyQ8ykFGZ4sl0t+KKxiuVzI5qlK5TCcn5WTKlzywpZcAAzEgJZQNbKAdkdSrIkFgJBqOxlNwEl4zkF5HcLkgg1aDnaHVKZX85VC3dYPUgvRAfXg/ZIpP0hiNxtNZgslgLfI95GgbecApc7lcy0FYgALWZ8q4mhWRvS2+3quNa9tK90RT3emm7HsuACiUAgwVmUmAZDA1aQAHFqMAZ3OF0uwDYU0DQeDgDTp7P52BF8vCDaiButxfd8uDyOkEeMafM0A==)

**Comment by UpperAvenue** [comment]
@svick I see what you're saying now.  You're right, we wouldn't able to resolve the method at compile time in your example.  Thanks for taking the time!
